Workflow for Churchill Lab's g2gtools. Used for incorporating SNPs and Indels into reference genomes https://doi.org/10.1093/bioinformatics/bty078 - supplementary

Ex. Make strain-specific transcriptomes for each Diversity Outbred Founder from C57BL6NJ reference genome

* Ensembl public release 75 mus musculus: 
  * ftp://ftp.ensembl.org/pub/release-75/gtf/mus_musculus
* Indels and SNPs for GRCm38: 
  * ftp://ftp-mouse.sanger.ac.uk/REL-1410-SNPs_Indels/
* Genome Sequence: 
  * ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M20/GRCm38.p6.genome.fa.gz
* Genome Annotation:
  * ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M20/gencode.vM20.chr_patch_hapl_scaff.annotation.gtf.gz
* Indels vcf and vcf tbi REL-1410
  * vcf: ftp://ftp-mouse.sanger.ac.uk/REL-1410-SNPs_Indels/mgp.v4.indels.dbSNP.vcf.gz
  * tbi: ftp://ftp-mouse.sanger.ac.uk/REL-1410-SNPs_Indels/mgp.v4.indels.dbSNP.vcf.gz.tbi
* SNPs vcf and vcf tbi
  * vcf: ftp://ftp-mouse.sanger.ac.uk/REL-1410-SNPs_Indels/mgp.v4.snps.dbSNP.vcf.gz
  * tbi: ftp://ftp-mouse.sanger.ac.uk/REL-1410-SNPs_Indels/mgp.v4.snps.dbSNP.vcf.gz.tbi
